### ISSUE TITLE
UserSession.completeOAuth2 error in production React app

### ISSUE THREAD

**Issue by r-watkins** [Esri/arcgis-rest-js#580]
I get the following error when attempting to complete an authentication using UserSession.completeOAuth2:

```
TypeError: o.opener.parent[("__ESRI_REST_AUTH_HANDLER_" + z)] is not a function
    at r (UserSession.js:181)
    at Function.e.completeOAuth2 (UserSession.js:222)
    at auth.js:41
    at c (runtime.js:45)
    at Generator._invoke (runtime.js:264)
    at Generator.e.<computed> [as next] (runtime.js:98)
    at t (asyncToGenerator.js:3)
    at z (asyncToGenerator.js:25)
    at asyncToGenerator.js:32
    at new Promise (<anonymous>)
```

The app is built off of React (CRA) and is using the latest arcgis-rest packages (2.0.3).
It works with no errors in local but fails with this error when deployed to the production site. I am passing a URL parameter through the redirect which I'm not sure is affecting things (popup is set to false).

Even though it fails here and halts the app I can copy/paste the redirect URL that it produces into a new tab and the authentication goes through with no errors.

I'm stumped.

**Comment by tomwayson** [comment]
if you open the console of your built app does the `window` define a global variabl that starts w/ `__ESRI_REST_AUTH_HANDLER_`?

What version of react-scripts are you using?

**Comment by tomwayson** [comment]
FYI - this CRA (react-scripts@3.0.0) app uses arcgis-rest-js v2.0.3 and the log in works in prod mode:

https://create-arcgis-app.surge.sh/

Here's the source:

https://github.com/tomwayson/create-arcgis-app/

**Comment by r-watkins** [comment]
Thanks for the reply @tomwayson 

I don't see `__ESRI_REST_AUTH_HANDLER_` in the `window` object of either the production build that fails or the local build that works.

The project is currently on react-scripts 3.0.1.

In reference to the create-arcgis-app, it looks like it only needs to use the `UserSession.beginOAuth2` method because it has the popup enabled, is that right? I am trying to avoid the popup so I have to return from the redirect with `UserSession.completeOAuth2` where the error is occurring.

**Comment by tomwayson** [comment]
@patrickarlt 

I met w/ @r-watkins and I'm pretty sure this is a bug.

They are using non-popup OAuth flow, but they _launch_ the app from a link w/ `target="_blank"` and that causes `window.opener` to be defined. That in turn causes `completeOAuth2()` to enter the following block which looks like it's intended for popup flow:

https://github.com/Esri/arcgis-rest-js/blob/4d2975a0392488afff414f64784db12c0eea081c/packages/arcgis-rest-auth/src/UserSession.ts#L346-L353

As a fix I was thinking of something like:

```ts
      if (win.opener && win.opener.parent) {
        const handlerFn = win.opener.parent[`__ESRI_REST_AUTH_HANDLER_${clientId}`];
        if (handlerFn) {
          handlerFn(
            error ? JSON.stringify(error) : undefined,
            JSON.stringify(oauthInfo)
          );
          win.close();
          return undefined;
        }
      }
```

But wanted to run it by you first since you're more familiar than I am w/ what the exact scenario that block is supposed to handle.

**Comment by tomwayson** [comment]
@r-watkins

it's unlikely I'll have time to PR the above (w/ tests, etc) myself. If you or anyone else affected by this decides to, I can help shepherd the PR into being merged.

**Comment by tomwayson** [comment]
I'm going to consider this fixed btw #625 and #652. If anyone experiences this issue again, feel free to re-open.
